### ISSUE TITLE
Cancellation support for queueClient.ReceiveAsync()

### ISSUE THREAD

**Issue by ig-sinicyn** [Azure/azure-service-bus-dotnet#439]
WindowsAzure.ServiceBus v4.1.8

**Use case:** I have a code that receives and processes messages in a loop, something like this: 
```cs
		BrokeredMessage msg;
		while ((msg = await queueClient.ReceiveAsync(timeout)) != null)
		{
			// message processing goes here
		}
```
Messages are rare, timeout set to 2-5 minutes.

I want to be able to cancel receive op immediately on app shutdown, but there's no way to do that. The `Receive()` method does not accept CancellationToken.
I cannot abort receive via disposing the queue client, too. The client may be used by another parts of code and I do not want to abort it in the middle of message processing.
Use small timeouts? Nope, 1-sec receive timeout will result in a millions of billed operations on service bus (we have a lot of instances that poll the queue).

**Suggestion:** add cancellation support for the Receive() calls, please:)

**Comment by ig-sinicyn** [comment]
Ok, thanks!

**Comment by SeanFeldman** [comment]
@nemakam I'm running into this issue right now with the new client.
`messageReceiver.Receive()` by default has a timeout that and there's no way to indicate to the receivers to cancel their wait. Ideally, it should be possible to pass a cancellation token into `Receive()` operation and just to cancel that cancellation token to stop receiver's receive execution.

**Comment by nemakam** [comment]
@SeanFeldman  Doesn't calling `messageReceiver.CloseAsync()` stop the call?

**Comment by SeanFeldman** [comment]
@nemakam no, it doesn't 🙁
It waits for the operation timeout and then throws `ObjectDisposedException, Cannot access a disposed object.Object name: 'FaultTolerantAmqpObject`1'.

<details><summary>Exception stacktrace</summary>
at Microsoft.Azure.ServiceBus.Core.MessageReceiver.<OnReceiveAsync>d__87.MoveNext()--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)   at Microsoft.Azure.ServiceBus.Core.MessageReceiver.<>c__DisplayClass65_0.<<ReceiveAsync>b__0>d.MoveNext()--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)   at Microsoft.Azure.ServiceBus.RetryPolicy.<RunOperation>d__18.MoveNext()--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()   at Microsoft.Azure.ServiceBus.RetryPolicy.<RunOperation>d__18.MoveNext()--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)   at Microsoft.Azure.ServiceBus.Core.MessageReceiver.<ReceiveAsync>d__65.MoveNext()--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)   at Microsoft.Azure.ServiceBus.Core.MessageReceiver.<ReceiveAsync>d__63.MoveNext()--- End of stack trace from previous location where exception was thrown ---   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()   at UserQuery.<RunUserAuthoredQuery>d__0.MoveNext()
</details>

**Comment by SeanFeldman** [comment]
Another thing to mention @nemakam, is that closing a client should not be an option for cancelling current recieves. One might want to cancel the recieves, but keep the receiver to start receiving later (think of pause). I.e. receive operation should really take a cancellation token.

**Comment by nemakam** [comment]
@SeanFeldman 
I totally agree with the use case, but not really sure how easy it would be to implement this from this client perspective. The AMQP credit flow system has a potential to get into unexpected states if we cancel the operation. (Assuming we figured out a way to cancel it in the first place).

**Comment by SeanFeldman** [comment]
As a workaround, I'd close the receiver, but as noted above, that just doesn't work.
Meaning at any point in time when a client application needs to pause receiving or shutdown quickly, both operations are not possible.

**Comment by nemakam** [comment]
@SeanFeldman 
> It waits for the operation timeout and then throws ObjectDisposedException, Cannot access a disposed object.Object name: 'FaultTolerantAmqpObject1'.

When I call messageReceiver.CloseAsync(), the receive() call stops. I don't see what you are seeing

```csharp
var receiver = new MessageReceiver(TestUtility.NamespaceConnectionString, TestConstants.NonPartitionedQueueName, receiveMode: ReceiveMode.ReceiveAndDelete);

            try
            {
                // Ensuring link is created.
                var msg = await receiver.ReceiveAsync(TimeSpan.FromSeconds(5));
                Stopwatch watch = Stopwatch.StartNew();
                TaskExtensionHelper.Schedule(async () => {
                    await Task.Delay(TimeSpan.FromSeconds(10));
                    await receiver.CloseAsync();
                });
                await receiver.ReceiveAsync(TimeSpan.FromMinutes(1));
                watch.Stop();
                Assert.True(watch.Elapsed.TotalSeconds < 20);
            }
            finally
            {
                await receiver.CloseAsync().ConfigureAwait(false);
            }
```

**Comment by SeanFeldman** [comment]
@nemakam I have a feeling the line for "Ensuring link is created" is working around the bug with receive. If you comment it out, the receive operation will not be terminated earlier and exception will be thrown.

```c#
var receiver = new MessageReceiver(connectionString, "queue", ReceiveMode.ReceiveAndDelete);

// Ensuring link is created.
// WHEN this line is commented out, ObjectDisposedException is thrown
// var msg = await receiver.ReceiveAsync(TimeSpan.FromSeconds(5));

var receivingTask = receiver.ReceiveAsync(TimeSpan.FromMinutes(1));
await receiver.CloseAsync();

await receivingTask;
```

**Comment by nemakam** [comment]
Found a case when the CBS link is not being cancelled when the connection is aborted. Had a discussion with the AMQP library owner and it is being worked on right now.. Will release a fix for this soon

**Comment by SeanFeldman** [comment]
👍 @nemakam perhaps it should be a separate issue as it's not about cancellation token, but does fix a bug. I'll raise one.

**Comment by SeanFeldman** [comment]
Any updates on this issue? 
/cc @axisc

**Comment by nemakam** [comment]
With the other issue fixed in #527, I don't think this is still required.
If your module is closing down, and you need to stop, you could just close the `QueueClient`. That should stop it immediately. If you want to reuse this across multiple modules, then you can create a `ServiceBusConnection` and then create one dedicated `QueueClient` per module.
For Pause as well I would recommend recreating `QueueClient` using the same connection. Creating a link isn't expensive. The use case isn't worth re working on entire internals and potentially making the SDK more heavy to keep the credit flow system sane. I'm closing this issue as a wont-fix.

**Comment by SeanFeldman** [comment]
@nemakam I disagree with the above. Please re-open.

Here's a scenario where support for cancellation token is needed and closing a client is not an option.
A host is running with a certain concurrency level (`N`). This results in `N` number of concurrent receive operations and independent message processings. When the host is shutting down, I need to ensure the host is shutting down cleanly, i.e. messages that are in flight are processed and no new messages are retrieved anymore. This required to let messages in-flight to be completed while not to allow new ones to be received. Because `ReceiveAsync()` operation does not accept a cancellation token, it forces me to introduce unnecessary complexity to work around the problem.

In addition to that, passing `CancellationToken` into asynchronous APIs is encouraged and considered to be the right API design.

> In the Task Parallel Library (TPL), cancellation is not strongly tied to a cancel button, although that is one of the ways in which it frequently manifests itself in a UI. Cancellation can also be used in situations where we want to time out long-running tasks or cancel parallel tasks
>
> You should consider using cancellation for all methods that may take a long time to complete. While evaluating the execution time of an API, you should consider the worst-case scenario to inform your decision. ([.NET Blog, 2012](https://blogs.msdn.microsoft.com/dotnet/2012/06/06/async-in-4-5-enabling-progress-and-cancellation-in-async-apis/))

Bill Wagner in his More Effective C# [book](https://www.amazon.ca/More-Effective-Content-Update-Program/dp/0672337886) got it right in [item 33](http://www.informit.com/articles/article.aspx?p=2832590&seqNum=7)

> The task asynchronous programming model includes standard APIs for cancellation and reporting progress. These APIs are optional, but should be implemented correctly when the asynchronous work can effectively report progress or be cancelled.

It's a well established paradigm that developers expect and library developers should be providing.

> One of the benefits of async in C# is that tasks can be cancelled. This allows you to abort a task if user cancels it in the UI, navigates away from a web pages, etc. 
>
> That cancellation token, then, should be passed into any other asynchronous operations the method calls. It is the responsibility of the method to enable cancellation, if it can and desires to. Not all asynchronous tasks can be cancelled. Generally speaking, whether a task can be cancelled or not depends on whether the method has an overload that accepts CancellationToken.

MSDN article, [Async/Await - Best Practices in Asynchronous Programming](https://msdn.microsoft.com/en-us/magazine/jj991977.aspx) says:

> The next common problem is how to handle cancellation and progress reporting. The base class library (BCL) includes types specifically intended to solve these issues: `CancellationTokenSource`/`CancellationToken` and `IProgress<T>`/`Progress<T>`. Asynchronous code should use the Task-based Asynchronous Pattern, or TAP (msdn.microsoft.com/library/hh873175), which explains task creation, cancellation and progress reporting in detail.

| Problem                     | Solution                                                                 |
|------------------------|------------------------------------------------------|
| Support cancellation | CancellationTokenSource and CancellationToken |

where the referenced article, [Task-based Asynchronous Pattern](https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/task-based-asynchronous-pattern-tap#cancellation-optional) has the following:

> In TAP, cancellation is optional for both asynchronous method implementers and asynchronous method consumers. **If an operation allows cancellation, it exposes an overload of the asynchronous method that accepts a cancellation token** (CancellationToken instance). By convention, the parameter is named cancellationToken.

### To sum it up

1. Developers are trained to be able to cancel task-based operations by passing in `CancellationToken`.
1. True I/O operations (sending/receiving messages) are asynchronous and should be cancelled w/o forcing client to be closed.
1. When did ever Sean persisted for so long on something that wasn't a good idea? 🙂

**Comment by Sandiejat** [comment]
Seems like we aren't going to have this in Release 3.2.0! :(

**Comment by danielmarbach** [comment]
Given the number of upvotes on @SeanFeldman summary, why hasn't this be reopened @nemakam ?

**Comment by erikpaperik** [comment]
Pls reopen this issue.
We are currently facing the same issue as we are not able to gracefully shut down our receivers whilst processing already (yet uncompleted) messages.
